### ISSUE TITLE
Support N-Patch based UI elements

### ISSUE THREAD

**Issue by raysan5** [raysan5/raygui#82]
This idea is still under evaluation because it requires some raygui redesign. 

Current raygui implementation only supports some styling elements:
 - Borders Size
 - Colors (Border-Base-Text)
 - Text: Font, Size, Spacing
 - *Specific elements by control*

Despite I tried to avoid textures usage on ragui (a part of the font atlas), I'm considering adding support for a [n-patch texture](https://www.raylib.com/examples/web/textures/loader.html?name=textures_npatch_drawing) to define the style.

It would require reviewing all controls drawing, replacing the Border-Base rectangle-based calls by a custom `GuiDrawRectangle()` implementation that draws base-with-border or n-patch based.

Also, n-patch should be provided and configured properly... probably some API functions would be required for that.

Thinking about it...

**Comment by raysan5** [comment]
Not easy... every control should use the n-patch system in a different way...

**Comment by raysan5** [comment]
Note that all rectangles drawing has been centralized into:
```
GuiDrawRectangle(Rectangle rec, int borderWidth, Color borderColor, Color color);
```
This function takes care of drawing rectangle and border but every individual control takes care of calling this function with the corresponding values depending on the control state.

In case of an n-patch based styling system, things get more complicated because `GuiDrawRectangle()` should be aware of the control and state to choose the required n-patch rectangle accordingly. It would require a complete redesign of raygui to accomodate that change.

Out-of-scope for the moment, closing the issue.

**Comment by Demizdor** [comment]
@raysan5 Hello! 
Actually, i think you made a little easier to add such a thing.

I remember last year working on a way to skin the gui controls but never finished, partially because the `DrawTextureNPatch()` function provided by raylib doesn't tile the edges/center only stretches them so it was no good for me and that there were so many `DrawRectangle()` calls to check.

Ok, so i tried skinning a button on latest raygui and it worked pretty well, here's how i did it!
Firstly we need some structures holding the skin data for each control.

```c
//inside raygui.h
typedef struct {
    unsigned short x, y, width, height;         // source rect on texture
    unsigned short top, bottom, left, right;    // edges
} GuiSkinProperty;

typedef union {
    struct {
        GuiSkinProperty normal;
        GuiSkinProperty focused;
        GuiSkinProperty pressed;
        GuiSkinProperty disabled;
    };
    GuiSkinProperty state[4];
} GuiSkinStyle;

typedef GuiSkinProperty GuiSkinLabel;
typedef GuiSkinStyle GuiSkinButton;

typedef struct { 
    Texture tex;   // main skin texture
    bool useColor;  // use the style color
    
    GuiSkinLabel label;
    GuiSkinButton button;
    //... many more controls
} GuiSkin;
```

Now we need a global variable to store the current skin and some handy get/set functions. We use a pointer here because we want the end user to manage memory and also because we want to quickly change skins. `GuiDrawRectangle()` also needs to be changed a bit.
```c
//inside raygui.h
static GuiSkin* guiSkin = NULL;
// Sets the global skin
void GuiSetSkin(GuiSkin* skin)
{
    guiSkin = skin;
}
// Gets the global skin
GuiSkin* GuiGetSkin(void)
{
    return guiSkin;
}

static void GuiDrawRectangle(Rectangle rec, int borderWidth, Color borderColor, Color color, GuiSkinProperty* prop)
{
    if(prop != NULL && prop->width != 0 && prop->height != 0) 
    {
        // Draw 9patch https://en.wikipedia.org/wiki/9-slice_scaling
        
        // Use color overlay?
        const Color c = guiSkin->useColor ? color : RAYGUI_CLITERAL(Color){255, 255, 255, 255*guiAlpha};

        // NOTE: DrawTextureNPatch stretches instead of tilling :'(
         NPatchInfo npi = {RAYGUI_CLITERAL(Rectangle){prop->x, prop->y, prop->width, prop->height},prop->left,prop->top,prop->right,prop->bottom,NPT_9PATCH};
          DrawTextureNPatch(guiSkin->tex, npi, rec, RAYGUI_CLITERAL(Vector2){1.f,1.f}, 0.f, c);
    }
    else 
    {
        if (color.a > 0) 
        {
            // Draw rectangle filled with color
            DrawRectangle(rec.x, rec.y, rec.width, rec.height, color);
        }
        
        if (borderWidth > 0) 
        {
            // Draw rectangle border lines with color
            DrawRectangle(rec.x, rec.y, rec.width, borderWidth, borderColor);
            DrawRectangle(rec.x, rec.y + borderWidth, borderWidth, rec.height - 2*borderWidth, borderColor);
            DrawRectangle(rec.x + rec.width - borderWidth, rec.y + borderWidth, borderWidth, rec.height - 2*borderWidth, borderColor);
            DrawRectangle(rec.x, rec.y + rec.height - borderWidth, rec.width, borderWidth, borderColor);
        }
    }
}
```
`GuiButton()` would only have to be changed slightly (only 2 lines of code) and we're all set.
```c
//in raygui.h -> GuiButton()
    // Draw control
    //--------------------------------------------------------------------
    GuiSkinProperty* style = guiSkin != NULL && guiSkin->tex.id > 0 ? &guiSkin->button.state[state] : NULL;
    GuiDrawRectangle(bounds, GuiGetStyle(BUTTON, BORDER_WIDTH), Fade(GetColor(GuiGetStyle(BUTTON, BORDER + (state*3))), guiAlpha), Fade(GetColor(GuiGetStyle(BUTTON, BASE + (state*3))), guiAlpha), style);
```
Of course now we have to set the actual skin, that can be a little hard so on my previous try i made a skin editor ;) but now we have to do it manually. Here's a example.
```c
#include <stdio.h>
#define RAYGUI_IMPLEMENTATION
#include "raygui.h"
#define SCREEN_WIDTH 800
#define SCREEN_HEIGHT 450

int main(int argc, char **argv)
{
    InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, "Raygui Test");
    SetTargetFPS(60);
    
    // Setup skin
    GuiSkin skin = {0};
    skin.useColor = false;  // should the skin use the colors of current style
    skin.tex = LoadTexture("/backup/Tests/skin.png");
  
    // Setup button style
    GuiSkinButton button = {0};
    button.normal = (GuiSkinProperty){ .x=13, .y=13, .width=83, .height=23, .top=6, .left=6, .right=6, .bottom=6 };
    button.focused = (GuiSkinProperty){ .x=13, .y=49, .width=83, .height=23, .top=6, .left=6, .right=6, .bottom=6 };
    button.pressed = (GuiSkinProperty){ .x=13, .y=85, .width=83, .height=23, .top=6, .left=6, .right=6, .bottom=6 };
    button.disabled = (GuiSkinProperty){ .x=13, .y=121, .width=83, .height=23, .top=6, .left=6, .right=6, .bottom=6 };
    skin.button = button;
    
    GuiSetSkin(&skin);
    GuiLoadStyleDefault();
    GuiSetStyle(BUTTON, TEXT_COLOR_PRESSED, 0xfafafaff);
    GuiSetStyle(BUTTON, TEXT_COLOR_FOCUSED, 0xfafafaff);
    GuiFade(0.7f);
    
    while (!WindowShouldClose())
    {   
        BeginDrawing();
	ClearBackground(RAYWHITE);
        
       GuiGrid((Rectangle){0,0, SCREEN_WIDTH, SCREEN_HEIGHT}, 32.0f, 2);
       GuiButton((Rectangle){100, 100, 60, 110},  "CLICK ME!");
        
        EndDrawing();
    }
    CloseWindow();
    
    return 0;
}
```

![scrn](https://user-images.githubusercontent.com/27865535/85594067-46d19280-b650-11ea-93a5-a943d307b8a8.png)

And that's it, i think this approach could work, of course with the current way that `DrawTextureNPatch()` works this is just good to make only plain skins like rounded corners, for better skins that functions needs to be changed to somehow tile instead of streach. Do you want me to make a POC?



Oh and here is the "skin" texture i used 
![skin](https://user-images.githubusercontent.com/27865535/85594664-ccedd900-b650-11ea-8458-7d612f649806.png)

**Comment by raysan5** [comment]
Hi @Demizdor! Thanks for this super-detailed approach. I didn't consider using the `GuiSkin` and `GuiSkinStyle` structures.

Actually, I used a custom similar implementation for a small quizz videogame I made, that's the reason I considered adding it to raygui (and the reason to implement `GuiDrawRectangle()`. This is the n-patch image I used:

![npatch](https://user-images.githubusercontent.com/5766837/85615338-5e167d00-b65c-11ea-9579-8f89bc57f78d.png)

My tests worked very good for `GuiButton()` and some other control but that it does not work for every type of control (i.e. `GuiSlider()`, `GuiProgress()`, `GuiTextBox()`, `GuiListView()`...). Every control would require some specific implementation and trying to find a generic enough solution for everything would complicate code a lot. That's the reason I stopped it for now.

Finally I decided to keep `raygui` simple with limited styling.
